progetto_so/messageExchange/server.c

```diff
@@ -68,84 +49,146 @@ int main()
+
+		switch (cmd_detected[0]){
 			case '1':
 				clients_insert(pid);
 				break;
 			case '2':
 				p_clientsList = clients_display(n);
-				// printf("%s\n", p_clientsList);
-
-				// if(p_clientsList == NULL){
-				// 	// printf("%s\n", "no clients");
-				// 	sprintf(p_clientsList, "%s\n", "No Clients Connected");
-				// 	// strcpy(p_clientsList, "No Clients Connected\0");
-				// }
-		        
+				
 		        if(DEBUG)
 					printf("[DEBUG] %s\n", p_clientsList);
 
-				p_pipeName = (char*)malloc(strlen(PIPES_PATH)+strlen(pid)+1);
-
-				p_pipeName = concat(PIPES_PATH, p_pid);
-
-				strcpy(pipeName, p_pipeName); /* BANG!!! */
-
-				mknod(pipeName, S_IFIFO|0666, 0); /* Create named pipe */
-				
 				strcpy(clientsList, p_clientsList); /* BANG!!! */
 				
-				fd_client = open(pipeName, O_RDWR); /* Open it for writing */
-				// printf("fd: %d\n", fd_client);
-				write(fd_client, clientsList, sizeof(clientsList));
+				sendTextToClient(pid, clientsList);
 
-				// close(fd_client); /* Close pipe */
-				
 				if(DEBUG)
-					printf("[DEBUG] SIGUSR1 TO PID %d\n", i_pid);
+					printf("[DEBUG] SIGUSR1 TO PID %d\n", i_pid_m);
 
-				kill(i_pid, SIGUSR1);
+				//notify client through SIGNAL SIGUSR1
+				kill(i_pid_m, SIGUSR1);
 				break;		
 			case '3':
+
+				p_pid_d = strtok(NULL, " ");
+				p_msg = strtok(NULL, " ");
+
+				//TO RESOLVE THE CONCAT
+				// p_msg[0] = '\0';
+				// while(p_token = strtok(NULL, " "))
+				// 	strcat(p_msg,p_token);
+
+
+
 				if(DEBUG){
-					printf("[DEBUG] SENDING MESSAGE '%s' TO PID %d\n","", i_pid);
+					printf("[DEBUG] pid_d->%s\n", p_pid_d);
+					printf("[DEBUG] msg received: %s\n", p_msg);
+				}
+
+
+				if(clients_search(p_pid_d)==1){
+					//if the client is connected, send the message
+					i_pid_d = atoi(p_pid_d);			
+
+					sendTextToClient(p_pid_d, p_msg);
+					//notify client through SIGNAL SIGUSR1
+					kill(i_pid_d, SIGUSR1);
+				}else{
+					//otherwise notify the client who request to send this message
+					if(DEBUG)
+						printf("[DEBUG] Client %s is not connected.\n", p_pid_d);
+
+					//send SIGUSR2 to client
+					kill(i_pid_m, SIGUSR2);
+
 				}
-				printf("cmd: %s\n", cmd);
-				strtok_r(&p_pid[0], " ", &p_msg);
-				printf("msg: %s\n", p_msg);
 
-				// kill(i_pid, SIGUSR2);
 				break;
 			case '4':
 			case '5':
 				clients_delete(pid);
 				break;
-
 		}
+
+		printf("\n\r");
 	} 
 
 	if(DEBUG)
 		printf("%s\n", "End..");
 
-	close (fd); /* Close pipe */
-	unlink(CMD_PIPE_NAME); /* Remove used pipe */
+	close (i_fd); /* Close pipe */
+    remove(CMD_PIPE_NAME);
+
+}
+
+
+// init folder data to store pipes
+int initDataFolder() {
+	FILE *fp;
+
+	fp = popen("mkdir -p data/", "r");
+	if (fp == NULL) {
+		printf("[Error] - Error initialing process folder\n");
+		exit(1);
+	}
+	return 1;
+}
+
+//CTRL+C managing
+void sigHandler_1(int signumber) {
+    if(signumber == SIGINT){
+        remove(CMD_PIPE_NAME);
+        //kill clients?
+    	exit(0);
+    }
+    
```

Comment:
we have to insert a while to send SIGUSR3 to disconnect all clients to this server